File: Medium/11. Container With Most Water.java
class Solution {
    public int maxArea(int[] height) {
        int max = Integer.MIN_VALUE;
        int start = 0;
        int end = height.length - 1;
        while (start < end) {
            if ((end - start) * 10000 < max) {
                break;
            } else {
                if (height[start] > height[end]) {
                    max = Math.max(max, (end - start) * height[end]);
                    end--;
                } else {
                    max = Math.max(max, (end - start) * height[start]);
                    start++;
                }
            }
        }
        return max;
    }
}

File: Medium/11. Container With Most Water.py
from typing import List


class Solution:
    def maxArea(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        max_area = -float('inf')
        while l < r:
            if (r - l) * 10000 < max_area:
                break
            max_area = max(max_area, min(height[l], height[r]) * (r - l))
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        return max_area # type: ignore